gc/base/standard/Scavenger.cpp

```diff
@@ -4866,58 +4951,254 @@ MM_Scavenger::mutatorSetupForGC(MM_EnvironmentBase *envBase)
 #if defined(J9MODRON_TGC_PARALLEL_STATISTICS)
-			env->_scavengerStats._releaseScanListCount += 1;
+				targetEnv->_scavengerStats._releaseScanListCount += 1;
 #endif /* J9MODRON_TGC_PARALLEL_STATISTICS */
-			clearCache(env, env->_survivorCopyScanCache);
-			_scavengeCacheScanList.pushCache(env, env->_survivorCopyScanCache);
-			env->_survivorCopyScanCache = NULL;
+				addCacheEntryToScanListAndNotify(targetEnv, cache);
+			}
+			/* else we failed to push inactive cache since mutator thread reactivacted it -> concurrent cycle continues */
 		}
-		if (NULL != env->_deferredCopyCache) {
-			Assert_MM_true(env->_deferredCopyCache->flags & OMR_SCAVENGER_CACHE_TYPE_CLEARED);
-			Assert_MM_true(env->_deferredCopyCache->flags & OMR_SCAVENGER_CACHE_TYPE_COPY);
-			env->_deferredCopyCache->flags &= ~OMR_SCAVENGER_CACHE_TYPE_COPY;
+		/* else thread has had consecutive VM access release without any barriers in between to activate this copy cache */
+	}
+}
+
+MMINLINE void
+MM_Scavenger::handleDeferredCopyScanCache(MM_EnvironmentStandard *currentEnv, MM_EnvironmentStandard *targetEnv, bool flushCaches, bool final)
+{
+	MM_CopyScanCacheStandard *cache = targetEnv->_deferredCopyCache;
+	if (NULL != cache) {
+		/* We cannot allow dangling deferred copy cache - if we just pushed survivor or tenure copy cache we should push deferred, too. */
+		if ((currentEnv == targetEnv) || final || targetEnv->inNative()
+				|| (NULL == targetEnv->_survivorCopyScanCache) || (NULL == targetEnv->_tenureCopyScanCache)) {
+			/* Race between mutator thread actual owner and another mutator thread that helps with concurrent termination */
+			if ((uintptr_t)NULL != MM_AtomicOperations::lockCompareExchange((volatile uintptr_t *)&targetEnv->_deferredCopyCache, (uintptr_t)cache, (uintptr_t)NULL)) {
+				targetEnv->_deferredCopyCache = NULL;
+				Assert_MM_true(cache->flags & OMR_SCAVENGER_CACHE_TYPE_COPY);
+				Assert_MM_true(cache->flags & OMR_SCAVENGER_CACHE_TYPE_CLEARED);
+				if (flushCaches || (_waitingCount > 0) || (NULL == targetEnv->_survivorCopyScanCache) || (NULL == targetEnv->_tenureCopyScanCache)) {
+					/* Master thread in STW is MUTATOR type and can trigger this as well */
+					cache->flags &= ~OMR_SCAVENGER_CACHE_TYPE_COPY;
 #if defined(J9MODRON_TGC_PARALLEL_STATISTICS)
-			env->_scavengerStats._releaseScanListCount += 1;
+					targetEnv->_scavengerStats._releaseScanListCount += 1;
 #endif /* J9MODRON_TGC_PARALLEL_STATISTICS */
-			_scavengeCacheScanList.pushCache(env, env->_deferredCopyCache);
-			env->_deferredCopyCache = NULL;
+					addCacheEntryToScanListAndNotify(targetEnv, cache);
+				} else {
+					Assert_MM_true(NULL == targetEnv->_inactiveDeferredCopyCache);
+					targetEnv->_inactiveDeferredCopyCache = cache;
+				}
+			}
 		}
-		if (NULL != env->_tenureCopyScanCache) {
-			Assert_MM_true(env->_tenureCopyScanCache->flags & OMR_SCAVENGER_CACHE_TYPE_COPY);
-			env->_tenureCopyScanCache->flags &= ~OMR_SCAVENGER_CACHE_TYPE_COPY;
+	}
+}
+
+extern "C" {
+
```

Comment:
There is no need to have this code as c function, it can be member of `MM_Scavenger` class